File: src/main/java/com/turchinsky/entities/RouteDetailsEntity.java
package com.turchinsky.entities;


import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.*;
import java.util.Objects;

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Entity
@Table(name = "route_details")
@IdClass(RouteDetailsEntityPK.class)
public class RouteDetailsEntity {

    @Id
    @Column(name = "route_id", nullable = false, insertable = true, updatable = false)
    private int routeId;


    @Id
    @Column(name = "station_id", nullable = false, insertable = true, updatable = false)
    private int stationId;

    @Id
    @Column(name = "station_order", nullable = false)
    private int stationOrder;

    @Basic
    @Column(name = "km")
    private int km;

    @Basic
    @Column(name = "type")
    private int type;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "route_id", referencedColumnName = "id", insertable = false, updatable = false)
    private RouteEntity refRouteEntity;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "station_id", referencedColumnName = "id", insertable = false, updatable = false)
    private StationEntity refStationEntity;


    public RouteDetailsEntity(RouteEntity routeEntity, StationEntity stationEntity) {
        this.refRouteEntity = routeEntity;
        this.refStationEntity = stationEntity;
    }

    public RouteDetailsEntityPK createPK() {
        return new RouteDetailsEntityPK(this.getRouteId(), this.getStationId(),
                                        this.getStationOrder());
    }

    public boolean isPrimaryValuesNotZero(RouteDetailsEntity routeDetailsEntity) {
        return routeDetailsEntity.getRouteId() != 0 && routeDetailsEntity.getStationId() != 0 && routeDetailsEntity
                .getStationOrder() != 0;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        RouteDetailsEntity that = (RouteDetailsEntity) o;
        return getRouteId() == that.getRouteId() &&
                getStationId() == that.getStationId() &&
                getStationOrder() == that.getStationOrder() &&
                getKm() == that.getKm() &&
                getType() == that.getType();
    }

    @Override
    public int hashCode() {
        System.out.println(this.toString());
        return Objects.hash(getRouteId(), getStationId(), getStationOrder());
    }

    @Override
    public String toString() {
        return "RouteDetailsEntity{" +
                "routeId=" + routeId +
                ", stationId=" + stationId +
                ", stationOrder=" + stationOrder +
                ", km=" + km +
                ", type=" + type +
                '}';
    }

}

File: src/main/java/com/turchinsky/kt/RouteDetailsEntity.kt
//package com.turchinsky.kt
//
//import javax.persistence.*
//
//@Entity
//@Table(name = "route_details", schema = "public", catalog = "sbb")
//@IdClass(RouteDetailsEntityPK::class)
//open class RouteDetailsEntity {
//    @get:Id
//    @get:Column(name = "route_id", nullable = false, insertable = false, updatable = false)
//    var routeId: Int? = null
//
//    @get:Id
//    @get:Column(name = "station_id", nullable = false, insertable = false, updatable = false)
//    var stationId: Int? = null
//
//    @get:Id
//    @get:Column(name = "station_order", nullable = false, insertable = false, updatable = false)
//    var stationOrder: Int? = null
//
//    @get:Basic
//    @get:Column(name = "km", nullable = true)
//    var km: Int? = null
//
//    @get:Basic
//    @get:Column(name = "type", nullable = true)
//    var type: Int? = null
//
//    @get:ManyToOne(fetch = FetchType.LAZY)
//    @get:JoinColumn(name = "route_id", referencedColumnName = "id")
//    var refRouteEntity: RouteEntity? = null
//
//    @get:ManyToOne(fetch = FetchType.LAZY)
//    @get:JoinColumn(name = "station_id", referencedColumnName = "id")
//    var refStationEntity: StationEntity? = null
//
//    @get:OneToMany(mappedBy = "refRouteDetailsEntity")
//    var refScheduleDetailsEntities: List<ScheduleDetailsEntity>? = null
//
//    override fun toString(): String =
//            "Entity of type: ${javaClass.name} ( " +
//                    "routeId = $routeId " +
//                    "stationId = $stationId " +
//                    "stationOrder = $stationOrder " +
//                    "km = $km " +
//                    "type = $type " +
//                    ")"
//
//    // constant value returned to avoid entity inequality to itself before and after it's update/merge
//    override fun hashCode(): Int = 42
//
//    override fun equals(other: Any?): Boolean {
//        if (this === other) return true
//        if (javaClass != other?.javaClass) return false
//        other as RouteDetailsEntity
//
//        if (routeId != other.routeId) return false
//        if (stationId != other.stationId) return false
//        if (stationOrder != other.stationOrder) return false
//        if (km != other.km) return false
//        if (type != other.type) return false
//
//        return true
//    }
//
//}
//
//class RouteDetailsEntityPK : java.io.Serializable {
//    @get:Id
//    @get:Column(name = "route_id", nullable = false, insertable = false, updatable = false)
//    var routeId: Int? = null
//
//    @get:Id
//    @get:Column(name = "station_id", nullable = false, insertable = false, updatable = false)
//    var stationId: Int? = null
//
//    @get:Id
//    @get:Column(name = "station_order", nullable = false, insertable = false, updatable = false)
//    var stationOrder: Int? = null
//
//    override fun equals(other: Any?): Boolean {
//        if (this === other) return true
//        if (javaClass != other?.javaClass) return false
//        other as RouteDetailsEntityPK
//
//        if (routeId != other.routeId) return false
//        if (stationId != other.stationId) return false
//        if (stationOrder != other.stationOrder) return false
//
//        return true
//    }
//
//    // constant value returned to avoid entity inequality to itself before and after it's update/merge
//    override fun hashCode(): Int = 42
//
//}
